### PR TITLE
Make default IDP configurable, support OIDC IdPs

### LINKED ISSUE #12: [Feature Request] configurable IDP
**Is your feature request related to a problem? Please describe.**
Default Cognito(including user pool) is not available in some regions, such as `ap-east-1`, `cn-north-1` and `cn-northwest-1`.

Provide a configurable mechanism to enable other OIDC IDPs, such as KeyCloak.

**Describe the solution you'd like**
Use parameter to install/configure the supported IDP.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by dengmingtong**
Issue #, if available:

close #12 

Description of changes:

Checklist:

- [x] Have you successfully deployed to an AWS account with your changes?
- [x] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully tested with your changes locally?
- [x] If new dependencies have been added, have they been pinned to specific versions?
- [ ] Is this change also required on the AWS Solution version?
- [ ] Have you updated openapi.yaml if you made updates to API definition (including add, delete or update parameter and request data schema)?
- [ ] If you had to run manual tests, have you considered automating those tests by adding them to [end-to-end tests](../main/end-to-end-tests/README.md)?
- [ ] If you are updating the changelog and vending out a new release, have you updated versionNumber and versionDate in [.defaults.yml](../main/config/settings/.defaults.yml)

AS review ticket id:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/add-auth-providers.js [hunk 1]
```diff
@@ -13,151 +13,38 @@
...
 class AddAuthProviders extends Service {
   constructor() {
     super();
-    this.dependency([
-      'aws',
-      'authenticationProviderConfigService',
-      'authenticationProviderTypeService',
-      'cognitoUserPoolAuthenticationProvisionerService',
-    ]);
+    this.dependency(['oidcAddAuthProvider', 'cognitoAddAuthProvider']);
```
Comment (zxkane) [inline addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/add-auth-providers.js]: looks like those are name of services, could we make them as public constant?

File: addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/add-auth-providers.js [hunk 2]
```diff
@@ -13,151 +13,38 @@
...
+        await oidcAddAuthProvider.addAuthenticationProvider();
+        break;
+      }
+      case authProviderConstants.cognitoAuthProviderTypeId: {
+        const cognitoAddAuthProvider = await this.service('cognitoAddAuthProvider');
+        await cognitoAddAuthProvider.addAuthenticationProvider();
+        break;
       }
+      default:
+        console.log(`Sorry, we are out of ${defaultIdpType}.`);
```
Comment (zxkane) [inline addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/add-auth-providers.js]: I prefer to throw exception here. Missing correct auth type actually breaks the deployment.

File: addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/create-root-user-service.js [hunk 3]
```diff
@@ -13,198 +13,37 @@
...
-      } else {
-        throw this.boom.internalError(
-          `There was a problem getting the default native pool user. Username: ${nativeAdminUserEmail} UserPoolId: ${userPoolId}. Error code: ${err.code}`,
-          true,
-        );
+  async execute() {
+    const defaultIdpType = this.settings.get(settingKeys.defaultIdpType);
+    switch (defaultIdpType) {
+      case authProviderConstants.oidcAuthProviderTypeId: {
+        const oidcCreateRootUserService = await this.service('oidcCreateRootUserService');
```
Comment (zxkane) [inline addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/create-root-user-service.js]: I think we should merge the method `createRootUser`, `addAuthProvider` and others to the auth provider implementations? It's not necessary to create multiple services like `oidcXXX`, `oidcYYY`, `oidcZZZ`.

File: addons/addon-base-rest-api/packages/services/lib/authentication-providers/authentication-provider-config-service.js [hunk 4]
```diff
@@ -46,6 +46,21 @@ class AuthenticationProviderConfigService extends Service {
     return _.map(dbResults, toProviderConfig);
   }
 
+  async clearAllAuthenticationProviderConfigs() {
```
Comment (zxkane) [inline addons/addon-base-rest-api/packages/services/lib/authentication-providers/authentication-provider-config-service.js:49]: TODO

File: addons/addon-base-rest-api/packages/services/lib/authentication-providers/register-built-in-provider-services.js [hunk 5]
```diff
@@ -23,7 +25,9 @@ function registerBuiltInAuthProviders(container) {
     'cognitoUserPoolAuthenticationProviderService',
     new CognitoUserPoolAuthenticationProviderService(),
   );
-  container.register('userAttributesMapperService', new UserAttributesMapperService());
+  container.register('cognitoUserPoolUserAttributesMapperService', new CognitoUserPoolUserAttributesMapperService());
```
Comment (zxkane) [inline addons/addon-base-rest-api/packages/services/lib/authentication-providers/register-built-in-provider-services.js]: any attribute differences between `oidcUserAttributesMapperService` and `cognitoUserPoolUserAttributesMapperService`? should we extract an abstract of users for two implementations?

File: addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/oidc/create-root-user-service.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        // In case of any other error let it bubble up
+        throw this.boom.internalError(
+          `There was a problem creating the default native user in DDB. Username: ${adminUserEmail}. Error code: ${err.code}`,
+          true,
+        );
+      }
+    }
+  }
+
+  async createUser(rawData) {
```
Comment (zxkane) [inline addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/oidc/create-root-user-service.js]: This method `createUser` could be shared by all implementation. I prefer to share it in abstract class.

File: addons/addon-base-ui/packages/base-ui/src/models/authentication/Authentication.js [hunk 7]
```diff
@@ -90,11 +90,12 @@ const Authentication = types
       // Then store auth token as appIdToken on client
       let newIdToken;
       if (authCode) {
-        const mainUrl = document.location.href.split('?code')[0];
+        const mainUrl = document.location.href.split('?')[0];
```
Comment (zxkane) [inline addons/addon-base-ui/packages/base-ui/src/models/authentication/Authentication.js]: it'd better using url parser to get the parts you will use later

File: addons/addon-base-rest-api/packages/base-controllers/lib/__tests__/__fixtures__/authentication-provider-configs.js [hunk 8]
```diff
@@ -323,6 +323,7 @@ const configurations = [
 const publicConfigurations = [
```
Comment (zxkane) [inline addons/addon-base-rest-api/packages/base-controllers/lib/__tests__/__fixtures__/authentication-provider-configs.js:323]: I don't see any oidc auth configurations in test fixtures. We need add some for testing oidc auth provider.

File: addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/oidc/add-auth-provider.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    const authenticationProviderTypeService = await this.service('authenticationProviderTypeService');
+    const authenticationProviderTypes = await authenticationProviderTypeService.getAuthenticationProviderTypes(
+      getSystemRequestContext(),
+    );
+    const oidcAuthProviderTypeConfig = _.find(authenticationProviderTypes, {
+      type: authProviderConstants.oidcAuthProviderTypeId,
+    });
+
+    const authenticationProviderConfigService = await this.service('authenticationProviderConfigService');
+    await authenticationProviderConfigService.clearAllAuthenticationProviderConfigs();
```
Comment (zxkane) [inline addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/oidc/add-auth-provider.js:61]: why do we need clear the existing provider configurations here?

**Review by zxkane — CHANGES_REQUESTED** [review]

**Review by zxkane — COMMENTED** [review]
Some additional questions:

- is it enable to limit specific groups of OIDC to login
- how about the behaviors of built-in adding users feat when using OIDC to manage users externally?

File: addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/add-auth-providers.js (hunk 1 above)
Comment (dengmingtong) [reply addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/add-auth-providers.js]: Very good suggestion, the service names should be defined as public constant. Current situation is all the servcie names of the existing code are not defined as constant. So I suggest to keep current way. And we can change all services names to constant in another seperate commit and pr after we complete all function development tasks.

File: addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/add-auth-providers.js (hunk 2 above)
Comment (dengmingtong) [reply addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/add-auth-providers.js]: Good suggestion. Done

File: addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/create-root-user-service.js (hunk 3 above)
Comment (dengmingtong) [reply addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/create-root-user-service.js]: As we discuss offline, I will change code to registering the same name corresponds to different idp implementations during deployment process. When using these services, we don't need to distinguish which idp it is.

File: addons/addon-base-rest-api/packages/services/lib/authentication-providers/register-built-in-provider-services.js (hunk 5 above)
Comment (dengmingtong) [reply addons/addon-base-rest-api/packages/services/lib/authentication-providers/register-built-in-provider-services.js]: Good point, refactor code to register same service to different implementation according to idp type.

File: addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/oidc/create-root-user-service.js (hunk 6 above)
Comment (dengmingtong) [reply addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/oidc/create-root-user-service.js]: Good point, will refactor

File: addons/addon-base-ui/packages/base-ui/src/models/authentication/Authentication.js (hunk 7 above)
Comment (dengmingtong) [reply addons/addon-base-ui/packages/base-ui/src/models/authentication/Authentication.js]: good point, will refactor

File: addons/addon-base-rest-api/packages/base-controllers/lib/__tests__/__fixtures__/authentication-provider-configs.js (hunk 8 above)
Comment (dengmingtong) [reply addons/addon-base-rest-api/packages/base-controllers/lib/__tests__/__fixtures__/authentication-provider-configs.js:323]: Add oidc unit test

File: addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/oidc/add-auth-provider.js (hunk 9 above)
Comment (dengmingtong) [reply addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/oidc/add-auth-provider.js:61]: For cover the case that user change oidc idp, we need to clear existing config and create new.

File: addons/addon-base-raas/packages/.pnpm-debug.log [hunk 10]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (zxkane) [inline addons/addon-base-raas/packages/.pnpm-debug.log]: don't add those debugging log files to SCM, you might add it to `.gitignore` to exclude it

File: addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/cogito-user-pool/user-attributes-mapper-service.js [hunk 11]
```diff
@@ -127,14 +127,15 @@ class UserAttributesMapperService extends Service {
 
     if (username.includes('\\')) {
       usernameInIdp = _.split(username, '\\')[1];
-      username = username.replace('\\', '_');
+      // username = username.replace('\\', '_');
```
Comment (zxkane) [inline addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/cogito-user-pool/user-attributes-mapper-service.js]: directly remove it

**Review by zxkane — CHANGES_REQUESTED** [review]

File: addons/addon-base-raas/packages/.pnpm-debug.log (hunk 10 above)
Comment (dengmingtong) [reply addons/addon-base-raas/packages/.pnpm-debug.log]: Done.

File: addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/cogito-user-pool/user-attributes-mapper-service.js (hunk 11 above)
Comment (dengmingtong) [reply addons/addon-base-rest-api/packages/services/lib/authentication-providers/built-in-providers/cogito-user-pool/user-attributes-mapper-service.js]: Yes, remove them

**Review by zxkane — APPROVED** [review]

**Pull request merged**